### PR TITLE
Add support for multiple presets repositories to be active at once

### LINKED ISSUE #3437: Multiple Active Preset Sources
### Is your feature request related to a problem? Please describe

Only one preset source can be active at a time, so I cannot apply official presets and personal presets without switching preset sources.

### Describe the solution you'd like

I would like to store only my presets in my presets repo, then use the rest of the presets from the official repo. Currently I have to switch preset sources to do this. This request is for having multiple sources be active at once. So add all your sources, then pick presets from any of them. 

### Describe alternatives you've considered

I could fork the official repo and add my presets to that, but then I have to keep rebasing it to pull in new/updated official presets.

### Other information

KAACK

### PR DESCRIPTION
**Pull request by Benky**
### Description 
This adds support for multiple preset repositories to be active at once.

### Testing done
Locally with a few remote and local repos

### Links
Resolves #3437

### CODE REVIEW

File: src/tabs/presets/presets.js
Comment (haslinghuis) [inline src/tabs/presets/presets.js:504]:
```suggestion
        for (const preset of repo.index.presets) {
            if (this.isPresetFitSearch(preset, searchParams)) {
```

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
@limonspb can you test this :)

**Comment by limonspb** [comment]
This is awesome, thank you!
A couple of notes:
1. when multiple sources are selected, i think it still needs a warning (at the top of presets) that an unofficial source is selected.
2. Some preset source authors leave all official presets in the custom repos. That makes them show twice now. So now the "favorite presets functionality" must also remember, which preset source it was picked from. And don't show preset as a favorite for example, if it was picked from the official repo, but the current preset bar is from, let's say Chris Rosser's repo.
3. Also now i think when multiple sources are selected, the presets from official repo need a special icon indicating its from the official repo. For example like that:
image

**Comment by Benky** [comment]
@limonspb 
1) OK, that one should be easy 👍 
3) OK, I'll do that - good point! Do you want that logo to be in the bottom right or could I add it next to "favorite" icon, so we have all icons grouped together?
4) OK, I'll pop a message and log a warning about that - I was thinking about that, but kinda forgot to do that :-D
5 & 2) Thanks for mentioning these hashes - I can implement that deduplication, but thinking about the design a bit more, I think adding name of the preset source to these tiles (`PresetTitlePanelBody`) might help people to identify where each preset comes from (if that's important for them). I'll also modify "favorite presets functionality" as you're describing 👍

**Comment by limonspb** [comment]
> Thanks for comments and suggestions @limonspb!
> 
> * **1** - OK, that one should be easy 👍
> * **3** - OK, I'll do that - good point! Do you want that logo to be in the bottom right or could I add it next to "favorite" icon, so we have all icons grouped together?
> * **4** - OK, I'll pop a message and log a warning about that - I was thinking about that, but kinda forgot to do that :-D
> * **5 & 2** - Thanks for mentioning these hashes - I can implement that deduplication, but thinking about the design a bit more, I think adding name of the preset source to these tiles (`PresetTitlePanelBody`) might help people to identify where each preset comes from (if that's important for them). I'll also modify "favorite presets functionality" as you're describing 👍

thanks for doing that! i was procrastinating about this task for like a year now already lol

**Comment by Benky** [comment]
Hey @limonspb

Here's a screenshot of current version (build from the latest commit in this branch):

image

- warning for 3rd party repo should be visible when there's at least one 3rd party repo selected (visible on screenshot)
- I've implemented preset deduplication based on hash. We now store additional information in localStorage to be able to distinguish the same preset in different repos.
- new icon to distinguish official presets from 3rd party ones added (visible on screenshot)
- when repository fails to load, it's name is reported in the main screen (visible on screenshot)

**Comment by limonspb** [comment]
> Hey @limonspb
> 
> Here's a screenshot of current version (build from the latest commit in this branch):
> 
> image
> 
> * warning for 3rd party repo should be visible when there's at least one 3rd party repo selected (visible on screenshot)
> * I've implemented preset deduplication based on hash. We now store additional information in localStorage to be able to distinguish the same preset in different repos.
> * new icon to distinguish official presets from 3rd party ones added (visible on screenshot)
> * when repository fails to load, it's name is reported in the main screen (visible on screenshot)
> * there's a new row in preset overview (and detail) showing name of the repository the preset comes from

Awesome stuff, love the extra row with the source. Is it possible to hide it when ONLY betaflight official source is selected?
The majority of people, i assume, use only Betaflight repo. They don't need to see that its betaflight repo, unless they start messing with sources.

**Comment by Benky** [comment]
> > Hey @limonspb
> > Here's a screenshot of current version (build from the latest commit in this branch):
> > image
> > 
> > * warning for 3rd party repo should be visible when there's at least one 3rd party repo selected (visible on screenshot)
> > * I've implemented preset deduplication based on hash. We now store additional information in localStorage to be able to distinguish the same preset in different repos.
> > * new icon to distinguish official presets from 3rd party ones added (visible on screenshot)
> > * when repository fails to load, it's name is reported in the main screen (visible on screenshot)
> > * there's a new row in preset overview (and detail) showing name of the repository the preset comes from
> 
> Awesome stuff, love the extra row with the source. Is it possible to hide it when ONLY betaflight official source is selected? The majority of people, i assume, use only Betaflight repo. They don't need to see that its betaflight repo, unless they start messing with sources.
> 
> And was thinking to have BF icon at the bottom right corner, because some preset names are long, we probably don't want to cut them even more. What do you think? Cutting keywords shorter is ok, only a few people pay attention to them anyways :)

Hey @limonspb ,

cool idea regarding not showing it to ppl that have only official source selected. I've added that.

Regarding the icon - I've tried both and I kinda like it top-right a bit more - I like having these two icons together rather than scattered around. Also given that background of that icon is transparent, it shows most of the text it's overlayed over (see screenshot). There's not that many presets with name that long, so it's not something that would affect majority of the people - but as backend dev I know very little about UX, so your call :-))))

image

**Comment by limonspb** [comment]
you are the designer here ha :)
So if you think this one is better, let it be :)

A few notes about the code, much appreciated supporting the clean-small-function style, easy to read.

This section is being repeated in two files:
image

I think you can just move it into "PresetsRepoIndexed.js", because GithubRepo and WebsiteRepo both inherit from there.

**Comment by Benky** [comment]
Thanks a lot, @limonspb , I really appreciate the feedback! :)

I've fixed the code duplication, so let me know if there's more comments.

**Review by limonspb — APPROVED** [review]

File: src/tabs/presets/presets.js
Comment (haslinghuis) [inline src/tabs/presets/presets.js:120]: :space_invader:

**Review by haslinghuis — APPROVED** [review]

**Comment by haslinghuis** [comment]
Pulling the icons on the left side might be cleaner || limit title length.

image

**Comment by Benky** [comment]
> Pulling the icons to the left side might be cleaner || limit title length.
> 
> image

by "to the left side" you mean before the title itself?

**Comment by haslinghuis** [comment]
Yes can you try how that looks?

**Comment by Benky** [comment]
> Yes can you try how that looks - but trimming length could be an option?

I've tried having icons on the left and it looks weird to me (sorry...) - given that you can have mix of the non-BF and BF presets in one row / column, presets' title starts jumping around based on the amount of icons it's being prefixed with ...

I've implemented title trimming as suggested

File: src/tabs/presets/presets.js
Comment (Benky) [reply src/tabs/presets/presets.js]: fixed 👍

**Comment by nerdCopter** [comment]
i dont want to add work nor confusion, but an option could also be usage of sub-tabs much like the PIDS/Filters.
i.e. Official presets on a sub-tab.  source 2 presets on a sub-tab, source 3 presets on another tab....

**Comment by Benky** [comment]
@nerdCopter I was thinking about similar approach when I started working on this, but what made me stay with current visual implementation was filtering of the presets - when you'd looking for e.g. HDZero stuff, you have it aggregated in one page, compared to multiple tabs when we'd need to click through separate tabs per repo to find all filtered data.

I find it easier to have everything aggregated together (and distinguished by source repo tags / icon(s)) rather than forcing user click through 2-3 tabs to check if some of their favorite repos have some presets.

**Comment by limonspb** [comment]
> @nerdCopter I was thinking about similar approach when I started working on this, but what made me stay with current visual implementation was filtering of the presets - when you'd looking for e.g. HDZero stuff, you have it aggregated in one page, compared to multiple tabs when we'd need to click through separate tabs per repo to find all filtered data.
> 
> I find it easier to have everything aggregated together (and distinguished by source repo tags / icon(s)) rather than forcing user click through 2-3 tabs to check if some of their favorite repos have some presets.

yeah i agree that multiple tabs is not as nice. Better if search works in one place i think.

**Comment by sugaarK** [comment]
So you’ve trimed the titles? It didn’t look good having them over the logos. You could put it under the star other wise that would look cleaner

**Comment by sugaarK** [comment]
Or go through and edit the offenders to have a shorter title

**Comment by Benky** [comment]
> So you’ve trimed the titles? It didn’t look good having them over the logos. You could put it under the star other wise that would look cleaner

Yes, i'm trimming titles of presets, so icons (both logo and star) and text can no longer overlap.

**Comment by sugaarK** [comment]
We can set a character limit on titles

File: src/tabs/presets/TitlePanel/PresetTitlePanel.css
Comment (haslinghuis) [inline src/tabs/presets/TitlePanel/PresetTitlePanel.css:23]: @sugaarK this (I guess) trims the length of the title.

But agree we should set a limit in the preset itself.

Comment (haslinghuis) [reply src/tabs/presets/TitlePanel/PresetTitlePanel.css:23]: Anyway this should be ready to go

**Review by nerdCopter — APPROVED** [review]
* seems to function as intended; i did not do deep-testing

**Pull request merged**
